### PR TITLE
Add "device" hardware type

### PR DESCRIPTION
**Pull request by diurnalist**
This hardware type allows enrolling edge devices into the CHI@Edge v2.0 testbed. There are a few new workers:

- Tunelo: for provisioning channels via Tunelo, the tunneling/channel service.
- Balena: for configuring and registering the device on Balena
- Blazar (for devices): for enrolling the device in Blazar similar to hosts.

Tunelo is the 3rd worker to interface w/ a keystone service so I abstracted that mechanism.

⚠️ This includes a significant refactor of the Blazar module! The device plugin was almost identical to physical hosts (in particular the need to deal w/ availability windows), so I generalized into a BlazarBaseWorker. There is a DB migration as well to handle migrating the internal tracker of blazar resource state, this is what I am most concerned about re: the refactor.

⚠️ This includes a significant refactor of calling Keystone services! I moved them to a common helper construct so we don't have so much boilerplate.

### CODE REVIEW

**Comment by diurnalist** [comment]
[Doni "device" hardware type](https://trello.com/c/NOi4giei/2517-doni-device-hardware-type)

File: doni/driver/worker/blazar/__init__.py [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+            resource = call_blazar(
+                context,
+                self.resource_path,
+                method="post",
+                json=body,
+            ).get(self.resource_type)
+        except KeystoneServiceAPIError as exc:
+            if exc.code == 404:
+                # host isn't in ironic.
+                result["message"] = "Host does not exist in Ironic yet"
```
Comment (Mark-Powers) [inline doni/driver/worker/blazar/__init__.py]: This should belong in the physical_host implementation, or be a generic message

File: doni/driver/util.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    _msg_fmt = "%(service)s response malformed: %(text)s"
+
+
+def ks_service_requestor(
+    name,
+    client_factory: "Callable[[],Adapter]" = None,
+    microversion=None,
+    parse_error=None,
+) -> "Callable[[RequestContext, str, str, dict, list[int]], Union[Optional[dict],Optional[list]]]":
+    def _request(context, path, method="get", json=None, allowed_status_codes=[]):
```
Comment (Mark-Powers) [inline doni/driver/util.py]: `allowed_status_codes` seems a bit imprecise, since it allows <400 as well. Perhaps `allowed_error_codes`?

**Review by Mark-Powers — APPROVED** [review]
Overall, great work! The refactor looks sound. Just a few minor comments.

**Comment by diurnalist** [comment]
Thanks for the review!

**Pull request merged**